### ISSUE TITLE
Filter compile flags based on compile options.

### ISSUE THREAD

**Issue by erick-xanadu** [PennyLaneAI/catalyst#426]
Instead of having two pipelines, (one for async and one for non-async), we could have one single pipeline by predicating the compiler passes. For example:

```python
A_PIPELINE = (
  "A_PIPELINE_NAME",
  [
     ("a-pass", lambda options: return options.enable_a_pass), # conditionally added
     ("b-pass", lambda options: return True), # always added
  ]
```

And then we can just use a filter to get the final pass list.

**Comment by Hassan-Ali-17** [comment]
"Hi @erick-xanadu / @dime10, I'd like to work on this if it's still relevant. Since it's been open a while, I want to check — is the compile-flag/pipeline structure you described still accurate in the current codebase, or has this since been handled another way? I'll take a look at the frontend pipeline code in the meantime

**Comment by dime10** [comment]
Hi @Hassan-Ali-17, thank you for interest in Catalyst! I don't believe this issue is still relevant, perhaps there is something more recent that you find interesting? :)

**Comment by Hassan-Ali-17** [comment]
Yes perhaps there are some other issues that can be resolved by me and are
still open
